File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int minPrice = prices[0];
        int maxProfit = 0;

        for (int i = 1; i < prices.size(); ++i) {
            // Check profit if sold today
            int profit = prices[i] - minPrice;
            maxProfit = max(maxProfit, profit);
            // Update min price seen so far
            minPrice = min(minPrice, prices[i]);
        }

        return maxProfit;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int minSF = prices[0];
        int maxP = 0;

        for(int i=1; i<n;i++){
            int profit = prices[i] - minSF;

            maxP = Math.max(maxP, profit);

            minSF = Math.min(minSF, prices[i]);
        }
        return maxP;
    }
}